### PR TITLE
introduce net.Conn deadlines to avoid hung connections

### LINKED ISSUE #4133: unexpected hangs upon network errors in 'mc'
## Expected behavior
no hangs retries or fails to continue

## Actual behavior
```
{"name": "mc", "duration": "38601330", "function": "test_put_object_multipart", "status": "FAIL", "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-29667 --quiet --no-color cp /mint/data/datafile-65-MB mymini
o/mc-test-bucket-21981/mc-test-object-20287  >>> 
`/mint/data/datafile-65-MB` -> `myminio/mc-test-bucket-21981/mc-test-object-20287`
SIGABRT: abort
PC=0x46b7c1 m=0 sigcode=0

goroutine 0 [idle]:
runtime.fu
tex()
	runtime/sys_linux_amd64.s:552 +0x21
runtime.futexsleep(0x0?, 0x0?, 0x7fff68c91868?)
	runtime/os_linux.go:66 +0x36
runtime.notesleep(0x1a13368)
	runtime/lock_futex.go:159 +0x87
runtime.mPark(...)\
n	runtime/proc.go:1449
runtime.stopm()
	runtime/proc.go:2228 +0x8d
runtime.findrunnable()
	runtime/proc.go:2804 +0x865
runtime.schedule()
	runtime/proc.go:3187 +0x239
runtime.park_m(0xc000782340?)
	
runtime/proc.go:3336 +0x14d
runtime.mcall()
	runtime/asm_amd64.s:425 +0x43

goroutine 1 [select, 643 minutes]:
github.com/minio/mc/cmd.doCopySession({0x12c3bc8?, 0xc0007b6840}, 0xc0007ac6b0, 0xc0005196c0, 
0x0, 0xc0007ae1b0, 0x0)
	github.com/minio/mc/cmd/cp-main.go:568 +0x8b1
github.com/minio/mc/cmd.mainCopy(0xc0005196c0)
	github.com/minio/mc/cmd/cp-main.go:730 +0xd9c
github.com/minio/cli.HandleAction({0xf5d
100?, 0x117ffc8?}, 0x2?)
	github.com/minio/cli@v1.22.0/app.go:498 +0x5a
github.com/minio/cli.Command.Run({{0x10e9ba8, 0x2}, {0x0, 0x0}, {0x0, 0x0, 0x0}, {0x10f6cac, 0xc}, {0x0, ...}, ...}, ...)
	github.com/
minio/cli@v1.22.0/command.go:225 +0x9f7
github.com/minio/cli.(*App).Run(0xc000519860, {0xc00021c000, 0x8, 0x8})
	github.com/minio/cli@v1.22.0/app.go:261 +0x847
github.com/minio/mc/cmd.Main({0xc00021c000?, 0x
8, 0x8})
	github.com/minio/mc/cmd/main.go:136 +0x365
main.main()
	github.com/minio/mc/main.go:27 +0x2e

goroutine 34 [chan receive, 643 minutes]:
github.com/rjeczalik/notify.(*nonrecursiveTree).dispatch(
0xc0002bea80, 0x0?)
	github.com/rjeczalik/notify@v0.9.2/tree_nonrecursive.go:36 +0x45
created by github.com/rjeczalik/notify.newNonrecursiveTree
	github.com/rjeczalik/notify@v0.9.2/tree_nonrecursive.go:29 +
0x16a

goroutine 35 [chan receive, 643 minutes]:
github.com/rjeczalik/notify.(*nonrecursiveTree).internal(0xc0002bea80, 0xc0002bea20)
	github.com/rjeczalik/notify@v0.9.2/tree_nonrecursive.go:81 +0x5a
creat
ed by github.com/rjeczalik/notify.newNonrecursiveTree
	github.com/rjeczalik/notify@v0.9.2/tree_nonrecursive.go:30 +0x1be

goroutine 120 [chan receive, 643 minutes]:
github.com/minio/mc/cmd.trapSignals({0xc0
00608fa0, 0x3, 0x3})
	github.com/minio/mc/cmd/signals.go:36 +0xc5
created by github.com/minio/mc/cmd.Main
	github.com/minio/mc/cmd/main.go:133 +0x32f

goroutine 131 [syscall, 643 minutes]:
os/signal.sign
al_recv()
	runtime/sigqueue.go:151 +0x2f
os/signal.loop()
	os/signal/signal_unix.go:23 +0x19
created by os/signal.Notify.func1.1
	os/signal/signal.go:151 +0x2a

goroutine 123 [select]:
github.com/mini
o/mc/cmd.(*accounter).writer(0xc000594ba0)
	github.com/minio/mc/cmd/accounting-reader.go:73 +0x98
created by github.com/minio/mc/cmd.newAccounter
	github.com/minio/mc/cmd/accounting-reader.go:54 +0x10f

g
oroutine 151 [chan receive, 641 minutes]:
github.com/minio/minio-go/v7.(*Client).putObjectMultipartStreamFromReadAt(_, {_, _}, {_, _}, {_, _}, {_, _}, 0x4100000, ...)
	github.com/minio/minio-go/v7@v7.0.27/api
-put-object-streaming.go:214 +0xaa8
github.com/minio/minio-go/v7.(*Client).putObjectMultipartStream(_, {_, _}, {_, _}, {_, _}, {_, _}, 0x4100000, ...)
	github.com/minio/minio-go/v7@v7.0.27/api-put-object-stre
aming.go:49 +0x2bb
github.com/minio/minio-go/v7.(*Client).putObjectCommon(_, {_, _}, {_, _}, {_, _}, {_, _}, 0x4100000, ...)
	github.com/minio/minio-go/v7@v7.0.27/api-put-object.go:277 +0x718
github.com/mini
o/minio-go/v7.(*Client).PutObject(_, {_, _}, {_, _}, {_, _}, {_, _}, 0x4100000, ...)
	github.com/minio/minio-go/v7@v7.0.27/api-put-object.go:243 +0x245
github.com/minio/mc/cmd.(*S3Client).Put(0xc0006aa0e0, {0
x12c3bc8, 0xc0007b6840}, {0x12ba0e0, 0xc0001be0a0}, 0x4100000, {0x7f3bc22ed200, 0xc000594ba0}, {0xc0005aefc0, {0x0, ...}, ...})
	github.com/minio/mc/cmd/client-s3.go:1046 +0xb78
github.com/minio/mc/cmd.putTar
getStream({0x12c3bc8, 0xc0007b6840}, {0x7fff68c91da2, 0x7}, {0xc0006ac000, 0x41}, {0x0, 0x0}, {0x0, 0x0}, ...)
	github.com/minio/mc/cmd/common-methods.go:328 +0x25e
github.com/minio/mc/cmd.uploadSourceToTarge
tURL({0x12c3bc8, 0xc0007b6840}, {{0x0, 0x0}, 0xc00020de00, {0x7fff68c91da2, 0x7}, 0xc00003c600, 0x1, 0x0, ...}, ...)
	github.com/minio/mc/cmd/common-methods.go:584 +0x1adc
github.com/minio/mc/cmd.doCopy({0x12
c3bc8, 0xc0007b6840}, {{0x0, 0x0}, 0xc00020de00, {0x7fff68c91da2, 0x7}, 0xc00003c600, 0x1, 0x0, ...}, ...)
	github.com/minio/mc/cmd/cp-main.go:272 +0x40d
github.com/minio/mc/cmd.doCopySession.func3.3()
	git
hub.com/minio/mc/cmd/cp-main.go:555 +0x9f
github.com/minio/mc/cmd.(*ParallelManager).addWorker.func1()
	github.com/minio/mc/cmd/parallel-manager.go:105 +0x77
created by github.com/minio/mc/cmd.(*ParallelMana
ger).addWorker
	github.com/minio/mc/cmd/parallel-manager.go:94 +0x8a

goroutine 245 [select, 643 minutes]:
net/http.(*persistConn).roundTrip(0xc000424b40, 0xc00021a500)
	net/http/transport.go:2620 +0x974\
nnet/http.(*Transport).roundTrip(0xc0009fe3c0, 0xc0002ba600)
	net/http/transport.go:594 +0x7c9
net/http.(*Transport).RoundTrip(0x40d625?, 0x12b9dc0?)
	net/http/roundtrip.go:17 +0x19
net/http.send(0xc0002ba
600, {0x12b9dc0, 0xc0009fe3c0}, {0x10bfa60?, 0x1?, 0x0?})
	net/http/client.go:252 +0x5d8
net/http.(*Client).send(0xc0005af590, 0xc0002ba600, {0x0?, 0x0?, 0x0?})
	net/http/client.go:176 +0x9b
net/http.(*Cli
ent).do(0xc0005af590, 0xc0002ba600)
	net/http/client.go:725 +0x8f5
net/http.(*Client).Do(...)
	net/http/client.go:593
github.com/minio/minio-go/v7.(*Client).do(0xc0005b3290, 0xc000246493?)
	github.com/mi
nio/minio-go/v7@v7.0.27/api.go:490 +0xb3
github.com/minio/minio-go/v7.(*Client).executeMethod(0xc0005b3290, {0x12c3bc8, 0xc0007b6840}, {0x10ea11a, 0x3}, {0x0, {0xc000246493, 0x14}, {0xc0002464a8, 0x14}, ...})

	github.com/minio/minio-go/v7@v7.0.27/api.go:594 +0x655
github.com/minio/minio-go/v7.(*Client).uploadPart(0x0?, {0x12c3bc8, 0xc0007b6840}, {0xc000246493, 0x14}, {0xc0002464a8, 0x14}, {0xc00095a090, 0x24}, {0x1
2b8b40, ...}, ...)
	github.com/minio/minio-go/v7@v7.0.27/api-put-object-multipart.go:295 +0x445
github.com/minio/minio-go/v7.(*Client).putObjectMultipartStreamFromReadAt.func2(0x4, 0x0?)
	github.com/minio/m
inio-go/v7@v7.0.27/api-put-object-streaming.go:187 +0x696
created by github.com/minio/minio-go/v7.(*Client).putObjectMultipartStreamFromReadAt
	github.com/minio/minio-go/v7@v7.0.27/api-put-object-streaming.go
:158 +0x6b9

goroutine 134 [select, 643 minutes]:
net/http.(*persistConn).writeLoop(0xc000424b40)
	net/http/transport.go:2392 +0xf5
created by net/http.(*Transport).dialConn
	net/http/transport.go:1751 +
0x1791

goroutine 240 [semacquire, 643 minutes]:
sync.runtime_Semacquire(0xc00033d0e0?)
	runtime/sema.go:56 +0x25
sync.(*WaitGroup).Wait(0xc0007bc300?)
	sync/waitgroup.go:136 +0x52
github.com/minio/mc/c
md.(*ParallelManager).stopAndWait(...)
	github.com/minio/mc/cmd/parallel-manager.go:221
github.com/minio/mc/cmd.doCopySession.func3.1(...)
	github.com/minio/mc/cmd/cp-main.go:488
github.com/minio/mc/cmd.do
CopySession.func3()
	github.com/minio/mc/cmd/cp-main.go:499 +0x829
created by github.com/minio/mc/cmd.doCopySession
	github.com/minio/mc/cmd/cp-main.go:486 +0x7eb

goroutine 274 [select, 643 minutes]:
gi
thub.com/minio/minio-go/v7.(*Client).newRetryTimer.func2()
	github.com/minio/minio-go/v7@v7.0.27/retry.go:73 +0x145
created by github.com/minio/minio-go/v7.(*Client).newRetryTimer
	github.com/minio/minio-go
/v7@v7.0.27/retry.go:70 +0x16a

goroutine 133 [IO wait, 643 minutes]:
internal/poll.runtime_pollWait(0x7f3bc22bffc8, 0x72)
	runtime/netpoll.go:302 +0x89
internal/poll.(*pollDesc).wait(0xc0005c2380?, 0xc000
548000?, 0x0)
	internal/poll/fd_poll_runtime.go:83 +0x32
internal/poll.(*pollDesc).waitRead(...)
	internal/poll/fd_poll_runtime.go:88
internal/poll.(*FD).Read(0xc0005c2380, {0xc000548000, 0x1000, 0x1000})\
n	internal/poll/fd_unix.go:167 +0x25a
net.(*netFD).Read(0xc0005c2380, {0xc000548000?, 0x52f101?, 0xc0006c726c?})
	net/fd_posix.go:55 +0x29
net.(*conn).Read(0xc000596008, {0xc000548000?, 0x18?, 0xa9?})
	ne
t/net.go:183 +0x45
net/http.(*persistConn).Read(0xc000424b40, {0xc000548000?, 0x40563d?, 0x60?})
	net/http/transport.go:1929 +0x4e
bufio.(*Reader).fill(0xc0009fa000)
	bufio/bufio.go:106 +0x103
bufio.(*Rea
der).Peek(0xc0009fa000, 0x1)
	bufio/bufio.go:144 +0x5d
net/http.(*persistConn).readLoop(0xc000424b40)
	net/http/transport.go:2093 +0x1ac
created by net/http.(*Transport).dialConn
	net/http/transport.go:1
750 +0x173e

rax    0xca
rbx    0x0
rcx    0x46b7c3
rdx    0x0
rdi    0x1a13368
rsi    0x80
rbp    0x7fff68c91800
rsp    0x7fff68c917b8
r8     0x0
r9     0x0
r10    0x0
r11    0x286
r12    0x7fff68
c91810
r13    0x0
r14    0x1a12ae0
r15    0x0
rip    0x46b7c1
rflags 0x286
cs     0x33
fs     0x0
gs     0x0"}
```

## Steps to reproduce the behavior
Very hard and rare

## mc --version
latest release

## System information
containers - amd64

### PR DESCRIPTION
**Pull request by harshavardhana**
fixes #4133

### CODE REVIEW

File: cmd/flags.go [hunk 1]
```diff
@@ -48,6 +50,18 @@ var globalFlags = []cli.Flag{
 		Name:  "insecure",
 		Usage: "disable SSL certificate verification",
 	},
+	cli.DurationFlag{
+		Name:   "conn-read-deadline",
+		Usage:  "custom connection READ deadline",
+		Hidden: true,
+		Value:  time.Minute,
```
[the remark this answers is not in the document]
Comment (klauspost) [reply cmd/flags.go:57]: Yeah. 5-10 minutes should be fine. The only important part is that it doesn't hang forever, but eventually times out.

[the remark this answers is not in the document]
Comment (harshavardhana) [reply cmd/flags.go:57]: Updated to 10mins

**Review by vadmeste — APPROVED** [review]
LGTM

**Review by klauspost — APPROVED** [review]
Let's try it out 👍🏼

**Pull request merged**